nomos-services/data-availability/sampling/src/backend/kzgrs.rs

```diff
@@ -1,2 +1,397 @@
+// std
+use std::collections::{BTreeSet, HashMap, HashSet};
+use std::fmt::Debug;
+use std::time::{Duration, Instant};
 
+// crates
+use rand::distributions::Standard;
+use rand::prelude::*;
+use tokio::time;
+use tokio::time::Interval;
 
+use kzgrs_backend::common::blob::DaBlob;
+
+//
+// internal
+use super::DaSamplingServiceBackend;
+use nomos_core::da::BlobId;
+use nomos_da_network_core::SubnetworkId;
+
+const NON_TRIGGERED_ERR: &str =
+    "We should not receive a sampling success from a non triggered blobId";
+
+#[derive(Clone)]
+pub struct SamplingContext {
+    subnets: HashSet<SubnetworkId>,
+    started: Instant,
+}
+
+#[derive(Debug, Clone)]
+pub struct KzgrsDaSamplerSettings {
+    pub num_samples: u16,
+    pub old_blobs_check_interval: Duration,
+    pub blobs_validity_duration: Duration,
+}
+
+pub struct KzgrsDaSampler<R: Rng> {
+    settings: KzgrsDaSamplerSettings,
+    validated_blobs: BTreeSet<BlobId>,
+    pending_sampling_blobs: HashMap<BlobId, SamplingContext>,
+    rng: R,
+}
+
+impl<R: Rng> KzgrsDaSampler<R> {
+    fn prune_by_time(&mut self) {
+        self.pending_sampling_blobs.retain(|_blob_id, context| {
+            context.started.elapsed() < self.settings.blobs_validity_duration
+        });
+    }
+}
+
+// TODO there is no logging at all here. Should we not do some logging?
+// Or can we assume that it is done on an upeer level by the clients?
```

Comment:
We can add whatever logging we need here. Then it is gonna be wrapped in the service context.

Reply:
Added just 2 tracing::info